libnavigation-core/src/main/java/com/mapbox/navigation/core/history/MapboxHistoryReader.kt

```diff
@@ -0,0 +1,46 @@
+package com.mapbox.navigation.core.history
+
+import com.mapbox.navigation.core.history.model.HistoryEvent
+import com.mapbox.navigation.core.history.model.HistoryEventMapper
+import com.mapbox.navigator.HistoryReader
+import java.io.Closeable
+
+/**
+ * Allows you to read history files previously saved by [MapboxHistoryRecorder].
+ * All files in the [MapboxHistoryRecorder.fileDirectory] can be read with this reader.
+ *
+ * @param filePath absolute path to a file containing the native history file.
+ */
+class MapboxHistoryReader(
+    val filePath: String
+) : Closeable, Iterator<HistoryEvent> {
+
+    private val nativeHistoryReader = HistoryReader(filePath)
+    private val historyMapper = HistoryEventMapper()
+
+    private var next: HistoryEvent? = null
+
+    /**
+     * Returns `true` if the iteration has more elements.
+     */
+    override fun hasNext(): Boolean {
+        val historyRecord = nativeHistoryReader.next()
+        next = if (historyRecord != null) {
+            historyMapper.map(historyRecord)
+        } else {
+            null
+        }
+        return next != null
+    }
```

Comment:
This implementation of the `Iterator` proposed here doesn't behave as an iterator should.

Calling:
`hasNext`
`hasNext`
`next`

will not be equal to:
`hasNext`
`next`

that's why we shouldn't implement the `Iterator` interface.

Reply:
nice catch! Making it an `Iterator` gives a common interface and expectation (such as the bug you just found). This will make it easier to implement on top. The native interface does not provide a sdk friendly interface

Unfortunately mockk is having issues mocking native functions, have you seen this? https://github.com/mockk/mockk/issues/639